### PR TITLE
Added rule CanBeNonNullableProperty

### PR DESCRIPTION
**Pull request by severn-everett**
This addresses issue #4376

### CODE REVIEW

**Comment by severn-everett** [comment]
@BraisGabin The latest Detekt run on the branch had `CanBeNonNullableProperty` flag this call in `DetektExtension.kt`:
```
    @Deprecated("Use reportsDir which is equivalent", ReplaceWith("reportsDir"))
    val customReportsDir: File?
        get() = reportsDir
```
The problem is that Kotlin determines that `reportsDir` is non-nullable, even though it is Java code that could return a nullable. I'd say to have `CanBeNonNullableProperty` disqualify any property that derives its value from Java source code, but I'm not sure how I'd be able to test that in the Spec files. Any suggestions for how to compile Java code alongside Kotlin code in the tests, or should I just suppress the warning on `customReportsDir`?

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+ *     }
+ * }
+ *
+ * class A {
+ *     val a: Int
+ *         get() = 5
+ * }
+ * </compliant>
+ */
+@ActiveByDefault(since = "1.20.0")
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:68]: You need to add a `@RequiresTypeResolution` annotation here

Comment (cortinico) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:68]: Also I'm unsure we want to have this enabled by default

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+class CanBeNonNullableProperty(config: Config = Config.empty) : Rule(config) {
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Style,
+        "Property can be changed to non-nullable, as it is never set to null.",
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:73]:
```suggestion
        "Property can be changed to non-nullable, as it is never sets to null.",
```

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        Debt.TEN_MINS
+    )
+
+    override fun visitKtFile(file: KtFile) {
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:77]: You're missing a `super.visitKtFile(file)` call here

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            super.visitKtFile(file)
+            // Any candidate properties that were not removed during the inspection
+            // of the Kotlin file were never assigned nullable values in the code,
+            // thus they can be converted to non-nullable.
+            candidateProps.forEach { (_, property) ->
+                report(
+                    CodeSmell(
+                        issue,
+                        Entity.from(property),
+                        "A nullable property can be made non-nullable."
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:100]: nit: You could improve this to mention the name of the property as you have it

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullablePropertySpec.kt [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+                    fun foo(): Int {
+                        val b = a
+                        a = b + 1
+                        fileA = a + 1
+                        val a = null
+                        return b
+                    }
+                }
+                """
+            Assertions.assertThat(subject.compileAndLintWithContext(env, code)).hasSize(5)
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullablePropertySpec.kt:60]: Can you split into smaller tests?

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullablePropertySpec.kt [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+                        5
+                    }
+                    
+                    init {
+                        b = 5
+                        c = cVal
+                    }
+                }
+                """
+            Assertions.assertThat(subject.compileAndLintWithContext(env, code)).hasSize(5)
```
Comment (cortinico) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullablePropertySpec.kt:175]: Same as above

**Review by cortinico — COMMENTED** [review]
> I'd say to have CanBeNonNullableProperty disqualify any property that derives its value from Java source code

+1 on this

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt (hunk 2 above)
Comment (severn-everett) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:73]: The sentence is grammatically correct as it is; `sets` would work only if the word `is` is removed

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt (hunk 3 above)
Comment (severn-everett) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:77]: Why would this be necessary? The `NonNullableCheckVisitor` instance conducts the actual check of the file for the rule, not `CanBeNonNullableProperty` itself.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt (hunk 2 above)
Comment (cortinico) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:73]: > The sentence is grammatically correct as it is; `sets` would work only if the word `is` is removed

Yup my bad, I was reading it too fast 👍

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt (hunk 3 above)
Comment (cortinico) [reply detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/CanBeNonNullableProperty.kt:77]: > Why would this be necessary?

That's part of the `Rule` contract and in general follows the visitor design pattern implementation. If by any chance we decide to implement a new feature on detekt that relies on the `visitKtFile` and lives inside the `Rule` top level class (or any of its ancestors), this rule will break this behavior.

**Comment by BraisGabin** [comment]
>  but I'm not sure how I'd be able to test that in the Spec files. Any suggestions for how to compile Java code alongside Kotlin code in the tests, or should I just suppress the warning on `customReportsDir`?

@t-kameyama implemented the option to do that in #4203. I think that you can use the same pattern.

File: config/detekt/detekt.yml [hunk 7]
```diff
@@ -143,6 +143,8 @@ potential-bugs:
     active: true
 
 style:
+  CanBeNonNullableProperty:
+    active: false
```
Comment (BraisGabin) [inline config/detekt/detekt.yml:147]: I think that it's safe to enable this. Or, if it isn't, remove these lines

**Review by BraisGabin — COMMENTED** [review]
Run the tests with `-Pcompile-test-snippets=true` in the gradle command. That will make the tests run slower but it will ensure that your snippet codes compile (It is necessary to merge the PR)

This rule is really good! But I'm wondering if this same rule should check for the same pattern in local variables or should we use other rule. I think that this rule is ready to merge now but, if we want to extend it later with the local variables we should change its name before merging it. I mean, it's not needed to implement the local variable thing now but if we want both things in the same ruel it can't be called "property".

**Comment by severn-everett** [comment]
Should be good to go, but something happened on the Windows Java 8 build. It looks like Gradle crashed for whatever reason, not sure if it's got anything to do with the code.

**Review by BraisGabin — APPROVED** [review]

**Review by BraisGabin — APPROVED** [review]

**Comment by DHosseiny** [comment]
It seems No doc for this in detekt .dev

**Comment by cortinico** [comment]
> It seems No doc for this in detekt.dev

Doc is here: https://detekt.dev/docs/rules/style/#canbenonnullable

**Pull request merged**